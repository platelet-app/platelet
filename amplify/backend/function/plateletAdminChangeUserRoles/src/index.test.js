const indexModule = require("./index");
const aws = require("aws-sdk");

jest.mock(
    "/opt/appSyncRequest",
    () => {
        return {
            request: jest.fn(),
            errorCheck: jest.fn(),
        };
    },
    { virtual: true }
);
jest.mock(
    "/opt/graphql/mutations",
    () => {
        return {
            updateUser: "updateUserMutation",
        };
    },
    { virtual: true }
);
jest.mock(
    "/opt/graphql/queries",
    () => {
        return {
            getUser: "getUserQuery",
        };
    },
    { virtual: true }
);

// mock CognitoIdentityServiceProvider class
//
const cognitoIdentityServiceProvider = Object.getPrototypeOf(
    new aws.CognitoIdentityServiceProvider()
);

const appsyncModule = require("/opt/appSyncRequest");

const userId = "userId";

describe("index.js", () => {
    afterEach(() => jest.resetAllMocks());
    test("changing a user's roles", async () => {
        appsyncModule.request
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        getUser: {
                            id: userId,
                            username: "username",
                            _version: 1,
                            roles: ["USER"],
                        },
                    },
                }),
            })
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        updateUser: {
                            id: userId,
                            username: "username",
                            _version: 2,
                            roles: [],
                        },
                    },
                }),
            })
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        updateUser: {
                            username: "username",
                            id: userId,
                            _version: 2,
                            roles: ["ADMIN", "RIDER", "USER"],
                        },
                    },
                }),
            });

        jest.spyOn(
            cognitoIdentityServiceProvider,
            "adminAddUserToGroup"
        ).mockReturnValue({
            promise: () => Promise.resolve(),
        });
        jest.spyOn(
            cognitoIdentityServiceProvider,
            "adminRemoveUserFromGroup"
        ).mockReturnValue({
            promise: () => Promise.resolve(),
        });
        jest.spyOn(
            cognitoIdentityServiceProvider,
            "adminListGroupsForUser"
        ).mockReturnValue({
            promise: () =>
                Promise.resolve({
                    Groups: [
                        { GroupName: "ADMIN" },
                        { GroupName: "RIDER" },
                        { GroupName: "USER" },
                    ],
                }),
        });

        await indexModule.handler({
            arguments: { userId, roles: ["ADMIN", "RIDER", "USER"] },
        });
        expect(appsyncModule.request).toMatchSnapshot();
    });
    test("changing a user's roles cognito failure and clean up", async () => {
        appsyncModule.request
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        getUser: {
                            id: userId,
                            username: "username",
                            _version: 1,
                            roles: ["USER"],
                        },
                    },
                }),
            })
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        updateUser: {
                            id: userId,
                            username: "username",
                            _version: 2,
                            roles: [],
                        },
                    },
                }),
            })
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        updateUser: {
                            username: "username",
                            id: userId,
                            _version: 2,
                            roles: ["ADMIN", "RIDER", "USER"],
                        },
                    },
                }),
            })
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        updateUser: {
                            id: userId,
                            username: "username",
                            _version: 2,
                            roles: [],
                        },
                    },
                }),
            })
            .mockResolvedValueOnce({
                json: () => ({
                    data: {
                        updateUser: {
                            username: "username",
                            id: userId,
                            _version: 2,
                            roles: ["USER"],
                        },
                    },
                }),
            });

        const adminAddUserToGroupSpy = jest
            .spyOn(cognitoIdentityServiceProvider, "adminAddUserToGroup")
            .mockReturnValue({
                promise: () => Promise.resolve(),
            });
        const adminRemoveUserFromGroupSpy = jest
            .spyOn(cognitoIdentityServiceProvider, "adminRemoveUserFromGroup")
            .mockReturnValue({
                promise: () => Promise.resolve(),
            });
        const adminListGroupsForUserSpy = jest
            .spyOn(cognitoIdentityServiceProvider, "adminListGroupsForUser")
            .mockReturnValue({
                promise: () =>
                    Promise.resolve({
                        Groups: [{ GroupName: "RIDER" }, { GroupName: "USER" }],
                    }),
            })
            .mockReturnValue({
                promise: () =>
                    Promise.resolve({
                        Groups: [{ GroupName: "USER" }],
                    }),
            });

        await expect(
            indexModule.handler({
                arguments: { userId, roles: ["ADMIN", "RIDER", "USER"] },
            })
        ).rejects.toThrowErrorMatchingInlineSnapshot(
            `"Roles not updated correctly on cognito"`
        );
        expect(appsyncModule.request).toMatchSnapshot();
        expect(adminAddUserToGroupSpy).toMatchSnapshot();
        expect(adminRemoveUserFromGroupSpy).toMatchSnapshot();
        expect(adminListGroupsForUserSpy).toMatchSnapshot();
    });
});
