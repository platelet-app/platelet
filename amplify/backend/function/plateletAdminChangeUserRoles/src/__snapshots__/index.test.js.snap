// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.js changing a user's roles 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "query": "getUserQuery",
        "variables": Object {
          "id": "userId",
        },
      },
      "https://api.example.com/graphql",
    ],
    Array [
      Object {
        "query": "updateUserMutation",
        "variables": Object {
          "input": Object {
            "_version": 1,
            "id": "userId",
            "roles": Array [],
          },
        },
      },
      "https://api.example.com/graphql",
    ],
    Array [
      Object {
        "query": "updateUserMutation",
        "variables": Object {
          "input": Object {
            "_version": 2,
            "id": "userId",
            "roles": Array [
              "ADMIN",
              "RIDER",
              "USER",
            ],
          },
        },
      },
      "https://api.example.com/graphql",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`index.js changing a user's roles cognito failure and clean up 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "query": "getUserQuery",
        "variables": Object {
          "id": "userId",
        },
      },
      "https://api.example.com/graphql",
    ],
    Array [
      Object {
        "query": "updateUserMutation",
        "variables": Object {
          "input": Object {
            "_version": 1,
            "id": "userId",
            "roles": Array [],
          },
        },
      },
      "https://api.example.com/graphql",
    ],
    Array [
      Object {
        "query": "updateUserMutation",
        "variables": Object {
          "input": Object {
            "_version": 2,
            "id": "userId",
            "roles": Array [
              "ADMIN",
              "RIDER",
              "USER",
            ],
          },
        },
      },
      "https://api.example.com/graphql",
    ],
    Array [
      Object {
        "query": "updateUserMutation",
        "variables": Object {
          "input": Object {
            "_version": 1,
            "id": "userId",
            "roles": Array [],
          },
        },
      },
      "https://api.example.com/graphql",
    ],
    Array [
      Object {
        "query": "updateUserMutation",
        "variables": Object {
          "input": Object {
            "_version": 2,
            "id": "userId",
            "roles": Array [
              "USER",
            ],
          },
        },
      },
      "https://api.example.com/graphql",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`index.js changing a user's roles cognito failure and clean up 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "GroupName": "ADMIN",
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
    Array [
      Object {
        "GroupName": "RIDER",
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
    Array [
      Object {
        "GroupName": "USER",
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
    Array [
      Object {
        "GroupName": "USER",
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
  ],
}
`;

exports[`index.js changing a user's roles cognito failure and clean up 4`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "GroupName": "COORDINATOR",
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
    Array [
      Object {
        "GroupName": "ADMIN",
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
    Array [
      Object {
        "GroupName": "COORDINATOR",
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
    Array [
      Object {
        "GroupName": "RIDER",
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
  ],
}
`;

exports[`index.js changing a user's roles cognito failure and clean up 5`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
    Array [
      Object {
        "UserPoolId": "somePoolId",
        "Username": "username",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "promise": [Function],
      },
    },
  ],
}
`;
